Assets/GothicVR/Scripts/World/DijkstraPathFinder.cs

```diff
@@ -0,0 +1,439 @@
+        {
+            if (Instance == null)
+            {
+                Instance = this;
+            }
+            else
+            {
+                Destroy(gameObject);
+            }
+        }
+
+        public void SetDijkstraWaypoints(Dictionary<string, DijkstraWaypoint> waypoints)
+        {
+            DijkstraWaypoints = waypoints;
+
+            foreach (var waypoint in DijkstraWaypoints)
+            {
+                var result = gameObject.FindChildRecursively(waypoint.Key).transform.position;
+                if (result != null)
+                {
+                    waypoint.Value.Position = result;
+                }
+            }
+            Debug.Log("DijkstraWaypoints set");
+            CalculateNeighbourDistances();
+
+        }
+
+        public DijkstraWaypoint[] FindFastestPath(string startWaypoint = null, string endWaypoint = null)
+        {
+            // Start the timer
+            var watch = System.Diagnostics.Stopwatch.StartNew(); 
+
+            // If start or end waypoints are not provided, use the default Start and End waypoints
+            if (startWaypoint == null || endWaypoint == null)
+            {
+                startWaypoint = Start;
+                endWaypoint = End;
+            }
+
+            // Get the start and end waypoints from the DijkstraWaypoints dictionary
+            var startDijkstraWaypoint = DijkstraWaypoints[startWaypoint];
+            var endDijkstraWaypoint = DijkstraWaypoints[endWaypoint];
+
+            // Initialize the previousNodes dictionary to keep track of the path
+            var previousNodes = new Dictionary<string, DijkstraWaypoint>();
+            // Initialize the unvisited priority queue to keep track of waypoints to be visited
+            var unvisited = new PriorityQueue();
+
+            // For each waypoint in DijkstraWaypoints 
+            foreach (var waypointx in DijkstraWaypoints.Values)
+            {
+                // If the waypoint is the start waypoint, set its SummedDistance to 0
+                if (waypointx.Name == startWaypoint)
+                {
+                    waypointx.SummedDistance = 0;
+                }
+                // Otherwise, set its SummedDistance to infinity
+                else
+                {
+                    waypointx.SummedDistance = double.MaxValue;
+                }
+
+                // Add the waypoint to the unvisited set and set its previous node to null
+                unvisited.Enqueue(waypointx, waypointx.SummedDistance);
+                // Add the waypoint to the previousNodes dictionary and set its previous node to null
+                previousNodes[waypointx.Name] = null;
+            }
+
+            while (unvisited.Count > 0)
+            {
+                var currentWaypoint = unvisited.Dequeue();
+                Debug.Log(currentWaypoint.Name + " " + currentWaypoint.SummedDistance);
+                LightUpWaypoint(currentWaypoint.Name, Color.yellow);
```

Comment:
I'd recommend you put these debug elements:
* remove them or
* hide them behind a feature flag

Your path finding algorithm is performance critical. Every last CPU cycle we can save, we should do so. ;-)

Reply:
Updated.